proposals/blittable.md

```diff
@@ -8,129 +8,92 @@
-
-A declarative, explicit opt in model makes it easier to code in this area.  Developers can be very explicit about the types they intend for interop.  This gives them compiler help when mistakes are made in their application and in any libraries they consume.  
-
-The lack of compile time support also makes it difficult to abstract over blittable types.  It's not possible for instance to author common helper routines using generic code:
+The primary motivation is to make it easier to author low level interop code in C#. Unmanaged types are one of the core building blocks for interop code, yet the lack of support in generics makes it impossible to create re-usable routines across all unmanaged types. Instead developers are forced to author the same boiler plate code for every unmanaged type in their library:
 
 ``` c#
-void Hash<T>(T value) where T : blittable struct
-{
-    using (T* p = &value) { 
-        ...
-    }
-}
+int Hash(Point point) { ... } 
+int Hash(TimeSpan timeSpan) { ... } 
 ```
 
-Instead developers are forced to rewrite virtually the same code for all of their blittable types:
+To enable this type of scenario the language will be introducing a new constraint: unmanaged:
 
 ``` c#
-void Hash(Point p) { 
-    ...
-}
-
-void Hash(Time t) { 
+void Hash<T>(T value) where T : unmanaged
+{
     ...
 }
 ```
 
-The lack of constraints here also make it impossible to have efficient conversions between streams of data and more structured types.  This is an operation that is common in networking stacks and serialization layers:
+This constraint can only be met by types which fit into the unmanaged type definition in the C# language spec. Another way of looking at it is that a type satisfies the unmanaged constraint iff it can also be used as a pointer. 
 
 ``` c#
-Span<byte> Convert<T>(ref T value) where T : blittable {
-    ...
-}
+Hash(new Point()); // Okay 
+Hash(42); // Okay
+Hash("hello") // Error: Type string does not satisfy the unmanaged constraint
 ```
 
-Such routines are advantageous because they are provably safe at compile time and allocation free.  Interop authors today can not due this (even though it's at a layer where perf is critical).  Instead they need to rely on allocating routines that have expensive runtime checks to verify values are correctly blittable.
-
-## Detailed design
-[design]: #detailed-design
+Type parameters with the unmanaged constraint can use all the features available to unmanaged types: pointers, fixed, etc ... 
 
-The language will introduce a new declaration modifier named `blittable` that can be applied to `struct` definitions.  
 
 ``` c#
-blittable struct Point 
+void Hash<T>(T value) where T : unmanaged
 {
-    public int X;
-    public int Y;
+    // Okay
+    using (T* p = &value) { 
```

Comment:
fixed?

Reply:
yes, looks like this was meant to be  `fixed`

Reply:
Blarg. Will fix.